taichi/transforms/auto_diff.cpp

```diff
@@ -8,6 +8,80 @@
 #include <typeinfo>
 
 TLANG_NAMESPACE_BEGIN
+class IndependentBlocksJudger : public BasicStmtVisitor {
+ public:
+  using BasicStmtVisitor::visit;
+
+  void visit(LocalLoadStmt *stmt) override {
+    for (auto &lane : stmt->src.data) {
+      touched_allocas_.insert(lane.var->as<AllocaStmt>());
+    }
+  }
+
+  void visit(LocalStoreStmt *stmt) override {
+    touched_allocas_.insert(stmt->dest->as<AllocaStmt>());
+  }
+
+  void visit(AtomicOpStmt *stmt) override {
+    // We don't need to check the global atomics inisde the range for-loops
+    // because
+    // 1. If the range for-loop is innermost, they will be captured by
+    // MakeAdjoint anyway
+    // 2. If the range for-loop is not innermost, they will be processed by
+    // another IndependentBlocksJudger
+    if (is_inside_loop_)
+      return;
+    TI_ASSERT(stmt->dest->is<GlobalPtrStmt>());
+    for (const auto &node : stmt->dest->cast<GlobalPtrStmt>()->snodes.data) {
+      if (node->has_grad()) {
+        qualified_atomics_ = false;
+        break;
+      }
+    }
+  }
+
+  void visit(RangeForStmt *stmt) override {
+    inner_most_loop_ = false;
+    is_inside_loop_ = true;
+    stmt->body->accept(this);
+    is_inside_loop_ = false;
+  }
+
+  static bool run(IRNode *root) {
+    IndependentBlocksJudger Judger;
+    Block *block = root->as<Block>();
+    root->accept(&Judger);
+    for (const auto &alloca : Judger.touched_allocas_) {
+      // Test if the alloca belongs to the current block
+      bool belong_to_this_block = false;
+      for (auto b = alloca->parent; b; b = b->parent_block()) {
```

Comment:
IIUC this can be replaced by a std::any? (or even simpler, make it a std::all and merge the logic of `qualified_local_` below?

Reply:
Have made it simpler.